File: sdk/qbiz.rs
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde::{Deserialize, Serialize};
use hmac::{Hmac, Mac};
use sha2::Sha256;
use hex;

type HmacSha256 = Hmac<Sha256>;

/// QBiz API Client SDK for Rust.
/// Provides methods to interface with the QBiz QRIS Gateway Middleware.
#[derive(Debug, Clone)]
pub struct QBizClient {
    api_key: String,
    base_url: String,
    client: reqwest::Client,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct InvoiceItem {
    pub name: String,
    pub price: i64,
    pub quantity: i32,
}

#[derive(Serialize, Debug, Clone)]
pub struct CreateInvoiceParams {
    pub order_id: String,
    pub amount: i64,
    pub callback_url: Option<String>,
    pub redirect_url: Option<String>,
    pub merchant_id: Option<String>,
    pub customer_name: Option<String>,
    pub customer_email: Option<String>,
    pub customer_phone: Option<String>,
    pub items: Option<Vec<InvoiceItem>>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Invoice {
    pub id: String,
    pub order_id: String,
    pub amount: i64,
    pub total_amount: i64,
    pub status: String,
    pub qr_string: String,
    pub created_at: String,
    pub expired_at: String,
}

#[derive(Deserialize, Debug, Clone)]
struct CreateInvoiceResponse {
    pub success: bool,
    pub invoice: Option<Invoice>,
    pub error: Option<String>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct InvoiceStatusResponse {
    pub success: bool,
    pub id: String,
    pub order_id: String,
    pub amount: i64,
    pub total_amount: i64,
    pub status: String,
    pub paid_at: Option<String>,
}

impl QBizClient {
    /// Initialize a new QBizClient.
    pub fn new(api_key: &str, base_url: Option<&str>) -> Result<Self, &'static str> {
        if api_key.is_empty() {
            return Err("API Key is required to initialize QBizClient");
        }
        let url = base_url.unwrap_or("http://localhost:8000").trim_end_matches('/').to_string();
        Ok(Self {
            api_key: api_key.to_string(),
            base_url: url,
            client: reqwest::Client::new(),
        })
    }

    /// Create a new dynamic QRIS invoice.
    pub async fn create_invoice(&self, params: CreateInvoiceParams) -> Result<Invoice, String> {
        let url = format!("{}/api/v1/invoices", self.base_url);
        
        let mut headers = HeaderMap::new();
        let auth_val = format!("Bearer {}", self.api_key);
        headers.insert(AUTHORIZATION, HeaderValue::from_str(&auth_val).unwrap());
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));

        let response = self.client.post(&url)
            .headers(headers)
            .json(&params)
            .send()
            .await
            .map_err(|e| format!("Connection error: {}", e))?;

        let status = response.status();
        let body = response.json::<CreateInvoiceResponse>()
            .await
            .map_err(|e| format!("Failed to parse response: {}", e))?;

        if !status.is_success() || !body.success {
            return Err(body.error.unwrap_or_else(|| format!("HTTP Error {}", status)));
        }

        body.invoice.ok_or_else(|| "Invoice missing in response".to_string())
    }

    /// Fetch payment status of an invoice.
    pub async fn get_invoice_status(&self, invoice_id: &str) -> Result<InvoiceStatusResponse, String> {
        let url = format!("{}/api/v1/invoices/{}/status", self.base_url, urlencoding::encode(invoice_id));
        
        let response = self.client.get(&url)
            .header(CONTENT_TYPE, "application/json")
            .send()
            .await
            .map_err(|e| format!("Connection error: {}", e))?;

        let status = response.status();
        if !status.is_success() {
            return Err(format!("QBiz API Error: HTTP {}", status));
        }

        response.json::<InvoiceStatusResponse>()
            .await
            .map_err(|e| format!("Failed to parse response: {}", e))
    }

    /// Verify HMAC-SHA256 signature received in webhook headers.
    pub fn verify_webhook(&self, payload_raw_body: &str, signature_header: &str, webhook_secret: &str) -> bool {
        if signature_header.is_empty() || webhook_secret.is_empty() {
            return false;
        }

        let mut mac = match HmacSha256::new_from_slice(webhook_secret.as_bytes()) {
            Ok(m) => m,
            Err(_) => return false,
        };

        mac.update(payload_raw_body.as_bytes());
        let result = mac.finalize();
        let computed_hex = hex::encode(result.into_bytes());

        let sig_bytes = computed_hex.as_bytes();
        let header_bytes = signature_header.as_bytes();

        if sig_bytes.len() != header_bytes.len() {
            return false;
        }

        // Constant-time comparison to prevent timing attacks
        let mut equal = 0;
        for i in 0..sig_bytes.len() {
            equal |= sig_bytes[i] ^ header_bytes[i];
        }
        equal == 0
    }
}

File: sdk/qbiz.go
// Package qbiz provides an idiomatic Golang client for the QBiz Dynamic QRIS Gateway Hub.
// Zero external dependencies (uses standard library net/http and crypto/hmac).
package qbiz

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// Client represents a configured QBiz API client instance.
type Client struct {
	APIKey     string
	BaseURL    string
	HTTPClient *http.Client
}

// PurchaseItem defines purchase item parameters for an invoice.
type PurchaseItem struct {
	Name     string `json:"name"`
	Quantity int    `json:"quantity"`
	Price    int    `json:"price"`
}

// CreateInvoiceParams specifies payload options for creating a dynamic QRIS invoice.
type CreateInvoiceParams struct {
	OrderID       string         `json:"order_id"`
	Amount        int            `json:"amount"`
	CallbackURL   string         `json:"callback_url,omitempty"`
	RedirectURL   string         `json:"redirect_url,omitempty"`
	MerchantID    string         `json:"merchant_id,omitempty"`
	CustomerName  string         `json:"customer_name,omitempty"`
	CustomerEmail string         `json:"customer_email,omitempty"`
	CustomerPhone string         `json:"customer_phone,omitempty"`
	Items         []PurchaseItem `json:"items,omitempty"`
}

// Invoice represents a generated dynamic QRIS invoice details.
type Invoice struct {
	ID            string         `json:"id"`
	MerchantID    string         `json:"merchant_id"`
	OrderID       string         `json:"order_id"`
	BaseAmount    int            `json:"base_amount"`
	UniqueCode    int            `json:"unique_code"`
	TotalAmount   int            `json:"total_amount"`
	Status        string         `json:"status"`
	QRISPayload   string         `json:"qris_payload"`
	CheckoutURL   string         `json:"checkout_url"`
	RedirectURL   string         `json:"redirect_url"`
	CustomerName  string         `json:"customer_name"`
	CustomerEmail string         `json:"customer_email"`
	CustomerPhone string         `json:"customer_phone"`
	Items         []PurchaseItem `json:"items"`
	ExpiredAt     string         `json:"expired_at"`
	CreatedAt     string         `json:"created_at"`
}

// InvoiceResponse wraps standard invoice creation API responses.
type InvoiceResponse struct {
	Success bool     `json:"success"`
	Invoice *Invoice `json:"invoice,omitempty"`
	Error   string   `json:"error,omitempty"`
}

// InvoiceStatusResponse represents polling query status output.
type InvoiceStatusResponse struct {
	Status      string  `json:"status"`
	TotalAmount int     `json:"total_amount"`
	PaidAt      *string `json:"paid_at,omitempty"`
}

// WebhookPayload represents the standard payment.success webhook event data.
type WebhookPayload struct {
	Event       string `json:"event"`
	InvoiceID   string `json:"invoice_id"`
	OrderID     string `json:"order_id"`
	AmountPaid  int    `json:"amount_paid"`
	PaidAt      string `json:"paid_at"`
}

// NewClient initializes a new QBizClient instance.
func NewClient(apiKey string, baseURL ...string) *Client {
	url := "http://localhost:8000"
	if len(baseURL) > 0 && baseURL[0] != "" {
		url = strings.TrimRight(baseURL[0], "/")
	}

	return &Client{
		APIKey:  apiKey,
		BaseURL: url,
		HTTPClient: &http.Client{
			Timeout: 15 * time.Second,
		},
	}
}

// CreateInvoice generates a new dynamic QRIS invoice.
func (c *Client) CreateInvoice(ctx context.Context, params CreateInvoiceParams) (*Invoice, error) {
	if params.OrderID == "" {
		return nil, errors.New("order_id is required")
	}
	if params.Amount <= 0 {
		return nil, errors.New("amount must be greater than 0")
	}

	bodyJSON, err := json.Marshal(params)
	if err != nil {
		return nil, fmt.Errorf("failed marshaling request: %w", err)
	}

	endpoint := fmt.Sprintf("%s/api/v1/invoices", c.BaseURL)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewBuffer(bodyJSON))
	if err != nil {
		return nil, fmt.Errorf("failed creating request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.APIKey))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed reading response: %w", err)
	}

	var apiResp InvoiceResponse
	if err := json.Unmarshal(respBody, &apiResp); err != nil {
		return nil, fmt.Errorf("invalid json response (HTTP %d): %s", resp.StatusCode, string(respBody))
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 || !apiResp.Success {
		if apiResp.Error != "" {
			return nil, errors.New(apiResp.Error)
		}
		return nil, fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(respBody))
	}

	return apiResp.Invoice, nil
}

// GetInvoiceStatus queries the payment settlement state of a given invoice ID.
func (c *Client) GetInvoiceStatus(ctx context.Context, invoiceID string) (*InvoiceStatusResponse, error) {
	if invoiceID == "" {
		return nil, errors.New("invoice_id is required")
	}

	endpoint := fmt.Sprintf("%s/api/v1/invoices/%s/status", c.BaseURL, invoiceID)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("failed creating request: %w", err)
	}

	req.Header.Set("Accept", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed reading response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP error %d fetching status: %s", resp.StatusCode, string(respBody))
	}

	var statusResp InvoiceStatusResponse
	if err := json.Unmarshal(respBody, &statusResp); err != nil {
		return nil, fmt.Errorf("failed decoding status response: %w", err)
	}

	return &statusResp, nil
}

// VerifyWebhookSignature validates the authenticity of an incoming HMAC SHA-256 webhook signature.
func VerifyWebhookSignature(rawBody []byte, signature, secretKey string, timestamp int64, maxToleranceSeconds ...int64) bool {
	if len(rawBody) == 0 || signature == "" || secretKey == "" {
		return false
	}

	// 1. Anti-replay timestamp tolerance check (default 5 minutes / 300s)
	if timestamp > 0 {
		tolerance := int64(300)
		if len(maxToleranceSeconds) > 0 && maxToleranceSeconds[0] > 0 {
			tolerance = maxToleranceSeconds[0]
		}
		currentTime := time.Now().Unix()
		diff := currentTime - timestamp
		if diff < 0 {
			diff = -diff
		}
		if diff > tolerance {
			return false // Expired or future timestamp (potential replay attack)
		}
	}

	// 2. Compute HMAC SHA-256
	mac := hmac.New(sha256.New, []byte(secretKey))
	mac.Write(rawBody)
	expectedSignature := hex.EncodeToString(mac.Sum(nil))

	return hmac.Equal([]byte(strings.ToLower(signature)), []byte(strings.ToLower(expectedSignature)))
}
